// safe-types/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Err = exports.Ok = exports.Result = exports.None = exports.Some = exports.Option = void 0;
const option_1 = require("./option");
Object.defineProperty(exports, "Option", { enumerable: true, get: function () { return option_1.Option; } });
const { Some, None } = option_1.Option;
exports.Some = Some;
exports.None = None;
const result_1 = require("./result");
Object.defineProperty(exports, "Result", { enumerable: true, get: function () { return result_1.Result; } });
const { Ok, Err } = result_1.Result;
exports.Ok = Ok;
exports.Err = Err;
var task_1 = require("./task");
Object.defineProperty(exports, "Task", { enumerable: true, get: function () { return task_1.Task; } });
var variant_1 = require("./variant");
Object.defineProperty(exports, "OptionVariant", { enumerable: true, get: function () { return variant_1.OptionVariant; } });
Object.defineProperty(exports, "ResultVariant", { enumerable: true, get: function () { return variant_1.ResultVariant; } });
var utils_1 = require("./utils");
Object.defineProperty(exports, "is_void", { enumerable: true, get: function () { return utils_1.is_void; } });
Object.defineProperty(exports, "is_never", { enumerable: true, get: function () { return utils_1.is_never; } });
Object.defineProperty(exports, "expect_never", { enumerable: true, get: function () { return utils_1.expect_never; } });
var helpers_1 = require("./helpers");
Object.defineProperty(exports, "get_at_path", { enumerable: true, get: function () { return helpers_1.get_at_path; } });
Object.defineProperty(exports, "has_at_path", { enumerable: true, get: function () { return helpers_1.has_at_path; } });
Object.defineProperty(exports, "err_or_ok", { enumerable: true, get: function () { return helpers_1.err_or_ok; } });
